pkg/cache/v3/linear.go

```diff
@@ -241,82 +267,114 @@ func (cache *LinearCache) computeResourceChange(sub Subscription, ignoreReturned
 	return changedResources, removedResources, nil
 }
 
+func computeSotwStableVersion(versionMap map[string]string) string {
+	// To enforce a stable hash we need to have an ordered vision of the map.
+	// TODO(valerian-roche): review if using https://pkg.go.dev/github.com/mitchellh/hashstructure/v2 is possible to remove the need to sort.
+	keys := make([]string, 0, len(versionMap))
+	for key := range versionMap {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+
+	mapHasher := fnv.New64()
+
+	buffer := make([]byte, 0, 8)
+	itemHasher := fnv.New64()
+	for _, key := range keys {
+		buffer = buffer[:0]
+		itemHasher.Reset()
+		itemHasher.Write([]byte(key))
+		mapHasher.Write(itemHasher.Sum(buffer))
+		buffer = buffer[:0]
+		itemHasher.Reset()
+		itemHasher.Write([]byte(versionMap[key]))
+		mapHasher.Write(itemHasher.Sum(buffer))
+	}
+	buffer = buffer[:0]
+	return hex.EncodeToString(mapHasher.Sum(buffer))
+}
+
 func (cache *LinearCache) computeSotwResponse(watch ResponseWatch, ignoreReturnedResources bool) (*RawResponse, error) {
-	changedResources, removedResources, err := cache.computeResourceChange(watch.subscription, ignoreReturnedResources, false)
+	changedResources, removedResources, err := cache.computeResourceChange(watch.subscription, ignoreReturnedResources, cache.useStableVersionsInSotw)
 	if err != nil {
 		return nil, err
 	}
-
 	if len(changedResources) == 0 && len(removedResources) == 0 && !ignoreReturnedResources {
 		// Nothing changed.
 		return nil, nil
 	}
 
-	returnedVersions := make(map[string]string, len(watch.subscription.ReturnedResources()))
-	// Clone the current returned versions. The cache should not alter the subscription
-	for resourceName, version := range watch.subscription.ReturnedResources() {
-		returnedVersions[resourceName] = version
-	}
-
-	cacheVersion := cache.getVersion()
-	var resources []types.ResourceWithTTL
+	// In sotw the list of resources to actually return depends on:
+	//  - whether the type requires full-state in each reply (e.g. lds, cds).
```

Comment:
nit: it's not "e.g.", which means for example, it's a hardcoded exhaustive list in the spec.

```suggestion
	//  - whether the type requires full-state in each reply (lds an cds).
```

Reply:
Updated the comment